lib/os/sem.c

```diff
@@ -0,0 +1,113 @@
+/*
+ * Copyright (c) 2019 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <misc/sem.h>
+#include <syscall_handler.h>
+
+#ifdef CONFIG_USERSPACE
+int sys_sem_init(struct sys_sem *sem, unsigned int initial_count,
+		unsigned int limit)
+{
+	if (sem == NULL || limit == 0 ||
+	    initial_count > limit || limit > INT_MAX) {
+		return -EINVAL;
+	}
+
+	atomic_set(&sem->futex.val, initial_count);
+	sem->limit = limit;
+
+	return 0;
+}
+
+unsigned int sys_sem_count_get(struct sys_sem *sem)
+{
+	int value = atomic_get(&sem->futex.val);
+
+	return value > 0 ? value : 0;
+}
+
+int sys_sem_give(struct sys_sem *sem)
+{
+	int ret = 0;
+	int old_value;
+
+	do {
+		old_value = atomic_get(&sem->futex.val);
+		if (old_value >= sem->limit) {
+			break;
+		}
+	} while (atomic_cas(&sem->futex.val, old_value,
+			old_value < 0 ? 1 : old_value + 1) == 0);
+
+	if (old_value < 0) {
+		ret = k_futex_wake(&sem->futex, true);
+
+		if (ret > 0) {
+			return 0;
+		}
+	} else if (old_value >= sem->limit) {
+		return -EAGAIN;
+	}
+
+	return ret;
+}
+
+int sys_sem_take(struct sys_sem *sem, s32_t timeout)
+{
+	int ret = 0;
+	int old_value;
+
+	do {
```

Comment:
Style: Make this outer loop `while (ret == 0 || ret == -EAGAIN)`.  "do" loops are always more confusing, and by initializing ret above you've guaranteed the entry condition to the loop already.  No one needs to read to the bottom of your function to see what's going on.

Reply:
@andyross I think do { } while (); can save cpu cycles.